### PR TITLE
add genesis create-cardano command

### LINKED ISSUE #3798: [FR] - Fix genesis generation for new clusters
**Internal/External**
*Internal*   if an IOHK staff member.

**Area**

*Other*   Any other topic (Delegation, Ranking, ...).

**Describe the feature you'd like**

The current way of bootstrapping a new testnet requires a bunch of manual intervention where you generate a byron and shelley genesis, then convert byron keys to shelley and overwrite the shelley generated keys, then update hashes in genesis, etc... What I'd really like to see is a simple way to create a genesis for a new testnet that starts in byron and transitions to shelley. Essentially this means the k constants and other magic in byron should be auto-generated to match the shelley genesis spec, the genesis, delegate and byron utxo keys should be output in both shelley and byron formats and a flag for specifying how much to not allocate to the payment utxo keys indicates what the reserves are.

Something like this would work for my needs, although I'm open to suggestions.

```
cardano-cli genesis create --all-eras --genesis-dir genesis --gen-genesis-keys 1 --gen-byron-utxo-keys 1 --testnet-magic 42 --supply 45000000000000000 --reserve 0.3 --start-time  "$(date -u -d "today + 60 minutes" +'%Y-%m-%dT%H:%M:%SZ')" 
```

**Describe alternatives you've considered**
Writing my own genesis generation tool.

**Additional context / screenshots**

### PR DESCRIPTION
**Pull request by cleverca22**
* combines generation of byron and shelley eras genesis
* auto generates a config based on a node config template
* slot timing and start time parameters override template
* generates all keys in byron era and converts to shelley era

fixed #3798

### CODE REVIEW

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs [hunk 1]
```diff
@@ -369,6 +393,250 @@ runGenesisCreate (GenesisDir rootdir)
...
+      case find (isCertForSK sk) (Map.elems $ dlgCertMap byronGenesis) of
+        Nothing -> throwE . NoGenesisDelegationForKey
+                   . Byron.prettyPublicKey $ getVerificationKey bSkey
+        Just x  -> pure x
+
+    dlgCertMap :: Genesis.GenesisData -> Map Byron.KeyHash Dlg.Certificate
+    dlgCertMap byronGenesis = Genesis.unGenesisDelegation $ Genesis.gdHeavyDelegation byronGenesis
+
+  --template <- readShelleyGenesisWithDefault (rootdir </> "genesis.spec.json") adjustTemplate
+  --alonzoGenesis <- readAlonzoGenesis (rootdir </> "genesis.alonzo.spec.json")
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs:637]: Why are these commented out?

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs [hunk 2]
```diff
@@ -369,6 +393,250 @@ runGenesisCreate (GenesisDir rootdir)
...
+                 -> Word  -- ^ num utxo keys to make
+                 -> Maybe SystemStart
+                 -> Maybe Lovelace
+                 -> BlockCount
+                 -> Word     -- slot length in ms
+                 -> Rational
+                 -> NetworkId
+                 -> FilePath -- Byron Genesis
+                 -> FilePath -- Shelley Genesis
+                 -> FilePath -- Alonzo Genesis
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs:472]: Should these comments use `-- ^`?

**Review by newhoggy — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs (hunk 2 above)
Comment (cleverca22) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs]: fixed

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs (hunk 1 above)
Comment (cleverca22) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs]: removed

File: cardano-api/src/Cardano/Api/Utils.hs [hunk 3]
```diff
@@ -50,3 +70,29 @@ failEither = either fail pure
 
 failEitherWith :: MonadFail m => (e -> String) -> Either e a -> m a
 failEitherWith f = either (fail . f) pure
+
+note :: MonadFail m => String -> Maybe a -> m a
+note msg = \case
+  Nothing -> fail msg
+  Just a -> pure a
+
+parseFilePath :: String -> String -> Opt.Parser FilePath
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Utils.hs:79]: All of these do not belong in Cardano.Api. They should live in cadano-cli.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs [hunk 4]
```diff
@@ -369,6 +393,246 @@ runGenesisCreate (GenesisDir rootdir)
...
+toVkeyJSON' = LBS.toStrict . textEnvelopeToJSON Nothing
+
+toOpCert :: (OperationalCertificate, OperationalCertificateIssueCounter) -> ByteString
+toOpCert = LBS.toStrict . textEnvelopeToJSON Nothing . fst
+
+toCounter :: (OperationalCertificate, OperationalCertificateIssueCounter) -> ByteString
+toCounter = LBS.toStrict . textEnvelopeToJSON Nothing . snd
+
+generateShelleyNodeSecrets :: [SigningKey GenesisDelegateExtendedKey] -> [VerificationKey GenesisKey]
+    -> IO (Map (Hash GenesisKey)
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs:413]: Wrong indent

File: cardano-api/src/Cardano/Api.hs [hunk 5]
```diff
@@ -663,7 +663,9 @@ module Cardano.Api (
 
     chainPointToSlotNo,
     chainPointToHeaderHash,
-    makeChainTip
+    makeChainTip,
```
[the remark this answers is not in the document]
Comment (cleverca22) [reply cardano-api/src/Cardano/Api.hs:666]: yeah, these could be moved until a cardano-cli utils file

File: cardano-api/src/Cardano/Api/Utils.hs (hunk 3 above)
Comment (cleverca22) [reply cardano-api/src/Cardano/Api/Utils.hs:79]: yeah, these could be moved until a cardano-cli utils file

**Pull request merged**